go/apps/bulk_message/tests/test_vumi_app.py

```diff
@@ -123,6 +122,268 @@ def test_consume_events(self):
+        batch_id = conversation.batch.key
+        yield self.app_helper.dispatch_command(
+            "bulk_send",
+            user_account_key=conversation.user_account.key,
+            conversation_key=conversation.key,
+            batch_id=batch_id,
+            dedupe=True,
+            content="hello world",
+            delivery_class="sms",
+            msg_options={},
+        )
+        yield self.app_helper.kick_delivery()
+        self.clock.advance(self.app.monitor_interval + 1)
+
+        contacts = yield self.get_opted_in_contacts(conversation)
+        msgs = yield self.app_helper.wait_for_dispatched_outbound(5)
+        contact_addrs = sorted([contact.msisdn for contact in contacts])
+        msg_addrs = sorted([msg["to_addr"] for msg in msgs])
+        self.assertNotEqual(contact_addrs, msg_addrs)
+        self.assertEqual(sorted(set(contact_addrs)), msg_addrs)
+
+    @inlineCallbacks
+    def test_interrupted_bulk_send_command_with_duplicates(self):
+        """
+        If we interrupt a bulk message command and reprocess it, we skip any
+        messages we have already sent.
+        """
+        # Replace send_message_via_window with one that we can break.
+        send_message_via_window = self.app.send_message_via_window
+        send_broken = DeferredQueue()
+        messages_sent = [0]
+
+        def breaking_send_message_via_window(*args, **kw):
+            if messages_sent[0] >= 4:
+                send_broken.put(None)
+                raise Exception("oops")
+            messages_sent[0] += 1
+            return send_message_via_window(*args, **kw)
+
+        self.app.send_message_via_window = breaking_send_message_via_window
+
+        group1 = yield self.app_helper.create_group_with_contacts(u'group1', 3)
+        group2 = yield self.app_helper.create_group_with_contacts(u'group2', 5)
+        conversation = yield self.app_helper.create_conversation(
+            groups=[group1, group2])
+        yield self.app_helper.start_conversation(conversation)
+        batch_id = conversation.batch.key
+
+        send_progress = yield self.app.get_send_progress(conversation)
+        self.assertEqual(send_progress, None)
+
+        command_params = dict(
+            user_account_key=conversation.user_account.key,
+            conversation_key=conversation.key,
+            batch_id=batch_id,
+            dedupe=False,
+            content="hello world",
+            delivery_class="sms",
+            msg_options={},
+        )
+        # We don't yield here, because this message will break the consumer and
+        # never be acked.
+        self.app_helper.dispatch_command("bulk_send", **command_params)
+        yield send_broken.get()
+        # Tell the fake broken that we've handled this message, because the
+        # now-broken consumer can't.
+        self.app.control_consumer._in_progress -= 1
+        self.app.control_consumer.channel.message_processed()
```

Comment:
Could we avoid these if we put them into a ``finally:`` clause in ``Consumer.consume``?